### DISCUSSION TITLE
Difference in JSON stored in Database for Repeatable Media Field (J3) vs. Subform Media Field (J4)

### DISCUSSION THREAD

**Discussion by justinherrin** [joomla/joomla-cms#39867] [General]
Hi all. I have a custom component built for J3 that I am in the process of updating to J4 and I have run into some trouble with the JSON values that are stored in the database when using the old [Repeatable Field](https://docs.joomla.org/Repeatable_form_field_type) -> Media Field in J3 vs. [Subform Field](https://docs.joomla.org/Subform_form_field_type) -> Media Field in J4.

The JSON stored is rather different from Repeatable Field in J3 to Subform Field in J4 and I am hoping someone can assist me with updating my JSON values to be compatible with J4.

Example of JSON stored with Repeatable Field -> Media Field in J3:
`{
	"book_image": [
		"images/books/cover.jpg",
		"images/books/inside.jpg",
		"images/books/back.jpg"
	]
}`

Example of JSON stored with Subform Field -> Media Field in J4:
`
{
	"book_images0": {
		"book_image":"images\/books\/cover.jpg#joomlaImage:\/\/local-images\/books\/cover.jpg?width=769&height=769"
	},
	"book_images1": {
		"book_image":"images\/books\/inside.jpg#joomlaImage:\/\/local-images\/books\/inside.jpg?width=769&height=769"
	},
	"book_images2": {
		"book_image":"images\/books\/back.jpg#joomlaImage:\/\/local-images\/books\/back.jpg?width=769&height=769"
	}
}
`

I have hundreds of records in the custom component stored, so any assistance to automate the process would be greatly appreciated.

Thanks in advance!

P.S. - Sorry the JSON code examples appear on single line. In the post I have everything in a more readable manner (line breaks, tabbed over, etc.) 🤷‍♂️

**Comment by richard67** [comment]
The repeatable fields should be converted automatically to subform fields when updating from 3.10 to 4.x with the Joomla Update Component, so why do you need to convert this manually?

**Comment by justinherrin** [comment]
@richard67 I updated from J3.10.x to 4.x locally using the Joomla Update Component and the update/updater did nothing for my custom extension. I wouldn't expect it to touch my custom extension files at all. I would have been quite surprised if it did anything.

**Reply by richard67** [reply]
I see. Was not clear to me from reading the initial post. I thought it's about the core custom fields, those are converted. Form fields used by 3rd party extensions of course are not.

**Reply by justinherrin** [reply]
Would you happen to be familiar with the code block(s) within the updater that took care of the conversions for core custom fields? Mainly I just need my JSON stored in the database updated, I already took care of changing my code to work with the subform field, but I have all these records in the database with outdated JSON formatting.

**Reply by richard67** [reply]
I know where it happens but I am not very familiar with the code: https://github.com/joomla/joomla-cms/blob/4.2-dev/administrator/components/com_admin/script.php#L221

**Reply by justinherrin** [reply]
Thanks for sharing that code block link. It's very well commented code, relatively easy to follow along with.

**Reply by richard67** [reply]
So your question has been answered? If so please mark the right answer. Thanks in advance.
